---
title: "SSP Lab Notebook"
author: "Sadie Allen"
date: "6/5/2017 - 8/11/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Week One (6/5/2017 - 6/9/2017)

### Background
Although originally my project was going to focus on aging in the kidneys, Gary chose to change the data set Taki and I were to investigate. Our new data, generated by the Alan D. Attie Lab of the University of Wisconsin-Madison, consists of phenotypical and mRNAseq information for 378 Diversity Outbred (DO) mice. The clinical phenotype data consists of characteristics including glucose, insulin, food consumption, and weight measurements. 

DO mice are different from most other laboratory mice as they are not inbred. Unlike C57BL/6J or A/J mice, which are completely genetically identical to allow for reproducible results when studying the effects of a particular environmental stressor, each DO mouse is genetically unique. DO mice are derived from eight founder strains (the same strains as the Collaborative Cross (CC) inbred strains). Randomized breeding resulted in genetically heterogeneous mice displaying a broad range of phenotypes. The genetic diversity of the DO mice provides several advantages for research. It greatly increases the power and precision of statistical methods, allows us to more accurately generalize conclusions to other genetically diverse populations, and allows us to examine the effects of genetic variation on known biological pathways and mechanisms. 

![Fig. 1: DO Mice](/Users/s-allens/Desktop/DO_mice.png)

All of the mRNAseq data provided in this data set comes from the pancreatic islet cells of the mice. In addition, all of the mice were fed a high fat diet (for a variable number of days) in order to emulate the progressive diet tendencies of humans in modern society. These factors, along with the glucose, insulin, and weight related phenotypes, optimize the Attie DO data set for use in studies of diabetes and obesity.

Diabetes is a metabolic disorder that inhibits the body's ability to maintain a healthy blood glucose level. Typically, people affected by the disease either cannot produce or become resistant to insulin, the hormone that allows glucose to enter cells and give them energy. Diabetes is a serious health concern in the United States. In 2010 diabetes was the seventh leading cause of death in the nation, and with 1.4 million Americans diagnosed every year, it needs to be addressed presently. 

### Accomplishments
#### Background Research 
During the first week of this project, we could not the data because we did not have accounts on the cadillac network where it was stored. I used the time to read several papers and familiarize myself with the topics and methods that would be relevant to my project: 

* The diversity outbred mouse population <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3524832/>
    + This paper familiarized me with the DO mouse strain's origin, advantages, and capabilities.
* Genetic background influences susceptibility to chemotherapy-induced hematoxicity <https://www.ncbi.nlm.nih.gov/pubmed/28607509>
    + Dan Gatti suggested this paper to Taki and I. It provided useful examples of how DO mice can be utilized in research. 
* Comprehensive alpha, beta and delta cell transcriptomes reveal that ghrelin selectively activates delta cells and promotes somatostatin release from pancreatic islets <http://www.sciencedirect.com/science/article/pii/S221287781630031X>
    + Gary provided this paper to us and was interested in the islet composition story. I learned about the functions and interactions of alpha, beta, and delta cells. This was also the starting point of my project (I will discuss this more later).
* Genetic architectures of quantitative variation in RNA editing pathways <https://www.ncbi.nlm.nih.gov/pubmed/26614740>
    + This paper provided another example of the use of DO mice in genetic research. 
    
#### Project Proposal 
I based my project heavily off of the findings in "Comprehensive alpha, beta and delta cell transcriptomes reveal that ghrelin selectively activates delta cells and promotes somatostatin release from pancreatic islets." The paper discussed ghrelin's regulatory role in insulin secretion of beta cells. The researchers found that ghrelin signals delta cells to secrete somatostatin, which in turn decreases the production of insulin by beta cells, providing a straigntforward explanation of the well-known insulinostatic actions of ghrelin. To start out, I decided to use the DO mice to investigate the effects of genetic diversity on the ghrelin-insulin pathway. The study hopefully would shed more light on the mechanisms that drive diabetes, informing future research and treatment possibilities. 

Note that the focus of my project throughout the project was fluid; I tried to keep fairly focused, but dead ends and new interest areas are inevitable in this kind of research. Unlike traditional science in which you pose a question and design an experiment to yield data specifically tailored to that question, I was given the task of foraging through an extensive data set for interesting findings. This kind of backward science lends itself to frequent focus shifts.

### Activities
#### High Performance Computing Workshop (6/8/2017)
In this workshop we learned the basics of coding in the Unix/Linux command line; navigating and working with files and directories, pipes and filters, loops, shell scripts, and automating a workflow. We also learned about connecting and interacting with clusters, the cadillac HPC queue system, running jobs on cadillac, and writing scripts to submit jobs to the cluster. 

## Week 2 (6/12/2017 - 6/16/2017)

### Accomplishments
#### Preliminary Phenotype Research
This week we were able to secure copy data from cadillac and begin preliminary research. 

The first step was to prepare the data for use in our analysis. The clinical phenotype data included measurements for 500 mice, but the mRNAseq data only had measurements for 378. I matched the mice by ID so I had only the 378 mice with both phenotype and mRNAseq data. 

Next, I had to check which phenotypes needed to be normalized, as many of the statistical methods I planned to use work significantly better on data following an approximate normal distribution. I made histograms of several phenotypes: 
```{r echo = FALSE, warning=FALSE, message = FALSE}
library(tidyverse)
library(plotly)
library(qtl2)
library(ggplot2)
library(dplyr)
library(corrplot)
library(reshape2)
matched_phenotypes <- read.csv("/Users/s-allens/Documents/ssp/summer_project/data/old_data/matched_pheno_clin.csv", as.is=TRUE)
ghrelin_shortlist <- matched_phenotypes %>% select(sex, Glu_0min,  Glu_sac, Ins_0min, 
                    Ins_sac, food_ave, weight_sac, G33_ins_secrete, 
                    G83_ins_secrete, G167_ins_secrete)
ghrelin_shortlist[,6:8] %>% 
  gather(var, value) %>%
  ggplot(aes(x=value)) +
  geom_histogram() +
  facet_wrap(~var, scales="free")
```

If they appeared to be heavily skewed, I log transformed them.

```{r echo = FALSE, warning=FALSE, message = FALSE}
library(tidyverse)
library(plotly)
library(qtl2)
library(ggplot2)
library(dplyr)
library(corrplot)
library(reshape2)
matched_phenotypes <- read.csv("/Users/s-allens/Documents/ssp/summer_project/data/old_data/matched_pheno_clin.csv", as.is=TRUE)
ghrelin_shortlist <- matched_phenotypes %>% select(sex, Glu_0min,  Glu_sac, Ins_0min, 
                    Ins_sac, food_ave, weight_sac, G33_ins_secrete, 
                    G83_ins_secrete, G167_ins_secrete)
ghrelin_shortlist[,6:8] %>% 
  gather(var, value) %>%
  ggplot(aes(x=log10(value))) +
  geom_histogram() +
  facet_wrap(~var, scales="free")
```

After normalizing the data and saving it to a new file, I ran some preliminary correlation scans between the phenotypes and gene expressions relevant to my research question.

```{r echo = FALSE, warning=FALSE, message = FALSE}
load("/Users/s-allens/Documents/ssp/summer_project/data/DO378_islet.RData")
source("scripts/functions.R")
library(corrplot)
matched_phenotypes <- read.csv("/Users/s-allens/Documents/ssp/summer_project/data/old_data/matched_pheno_clin.csv", as.is=TRUE)
ghrelin_shortlist <- matched_phenotypes %>% select(sex, Glu_0min,  Glu_sac, Ins_0min, 
                    Ins_sac, food_ave, weight_sac, G33_ins_secrete, 
                    G83_ins_secrete, G167_ins_secrete)
ghrl_exp <- get_exp_dat("Ghrl")
ghsr_exp <- get_exp_dat("Ghsr")
sst_exp <- get_exp_dat("Sst")
ins1_exp <- get_exp_dat("Ins1")
ins2_exp <- get_exp_dat("Ins2")
sstr3_exp <- get_exp_dat("Sstr3")
gene_expressions <- data.frame(ghrl = ghrl_exp, ghsr = ghsr_exp, 
                     sst = sst_exp, ins1 = ins1_exp, 
                     ins2 = ins2_exp, sstr = sstr3_exp)
ghrelin_shortlist <- cbind(ghrelin_shortlist, gene_expressions)
pcor <- cor(ghrelin_shortlist[2:16], use= "complete.obs")
#round(pcor, digits=2)
corrplot(pcor, order = "hclust" )
```

All of the correlations were fairly intuitive; G33_ins_secrete was highly correlated with G83_ins_secrete, Ins_0min was highly correlated with Ins_sac, etc. As expected, Ghsr (ghrelin receptor) and Sst (somatostatin) were negatively correlated with Ins1 and Ins2. 

### Activities
#### R Programming Workshop
I also attended a two day R programming workshop on Tuesday and Wednesday of this week. Since we had prior experience in R, Sue, who was teaching the course, appointed me, Taki, and Maddy to be helpers during the workshop. Although I already knew the basics of R, I learned a lot in this workshop and feel a lot more confident in my programming skills. Specifically, I have a much better understanding of subsetting and working with dataframses.

## Week 3 (6/19/2017 - 6/23/2017)

### Accomplishments
This was the first full week I had to explore the data set, since week one we did not have access to it and week two I had the R workshop Tuesday and Wednesday. I focused on two areas: generating p-value lists and learning how to use Karl Broman's QTL2 package. 

#### P-Value Lists
Gary suggested that we should loop through the 21,771 genes for which we have expression data and see which ones were significantly affected by sex using ANOVA and linear modeling. 
```{r warning= FALSE, message= FALSE}
source("scripts/functions.R")
load("/Users/s-allens/Documents/ssp/summer_project/data/DO378_islet.RData")
# Example: the effect of sex on Gnai3
gnai3_exp <- get_exp_dat("Gnai3")
anova(lm(gnai3_exp ~ annot.samples$Sex))
# Since the p-value is < 0.05, sex has a significant effect on Gnai3 islet expression.
```
I created a loop that ran through every gene and calculated the p-values. I then used this data to create a subset of genes that display a significant sex effect. 

Next, I applied the same statistical methods, but instead of looking at the effects of sex on gene expression, I chose 11 phenotypes (Glu_0min, Glu_sac, Ins_0min, food_ave, weight_sac, G33_ins_secrete, num_islets, Ins_per_islet, WPIC, and weight_change_ave) and ran through every gene for which we have expression data to determine which genes had significant effects on the phenotypes. 




























